### PR TITLE
Removing render.RenderableNode

### PR DESCRIPTION
**Pull request by paulbellamy**
Includes and supersedes #1169. Will need a rebase on master.

TODO:
- [x] Tests for detailed node rendering

### CODE REVIEW

**Comment by paulbellamy** [comment]
squashed and rebased up to 8c0fddc20d32c45687ddbd25e9001c17e065241b

File: report/counters.go [hunk 1]
```diff
@@ -88,21 +88,25 @@ func (c Counters) Merge(other Counters) Counters {
...
 
-func (c Counters) String() string {
-	if c.psMap == nil {
-		return "{}"
-	}
-	keys := []string{}
-	for _, k := range c.psMap.Keys() {
-		keys = append(keys, k)
+func (c Counters) ForEach(f func(key string, val int)) {
+	if c.psMap != nil {
```
Comment (tomwilkie) [inline report/counters.go]: Nit, but I'd prefer:

```go
if c.psMap == nil {
    return
}
...
```

File: report/counters.go [hunk 2]
```diff
@@ -88,21 +88,25 @@ func (c Counters) Merge(other Counters) Counters {
...
 	buf := bytes.NewBufferString("{")
-	for _, key := range keys {
-		val, _ := c.psMap.Lookup(key)
-		fmt.Fprintf(buf, "%s: %d, ", key, val)
-	}
+	prefix := ""
+	c.ForEach(func(k string, v int) {
+		fmt.Fprintf(buf, "%s%s: %d", prefix, k, v)
+		prefix = ", "
+	})
```
Comment (tomwilkie) [inline report/counters.go]: I like the prefix handling here.

File: render/detailed/metrics.go [hunk 3]
```diff
@@ -127,6 +143,11 @@ func (m *MetricRow) CodecDecodeSelf(decoder *codec.Decoder) {
 // NodeMetrics produces a table (to be consumed directly by the UI) based on
 // an origin ID, which is (optimistically) a node ID in one of our topologies.
```
Comment (tomwilkie) [inline render/detailed/metrics.go]: Comment out of date?

File: render/detailed/node.go [hunk 4]
```diff
@@ -184,19 +171,19 @@ var (
...
 
-func children(n render.RenderableNode) []NodeSummaryGroup {
+func children(n report.Node) []NodeSummaryGroup {
 	summaries := map[string][]NodeSummary{}
-	n.Children.ForEach(func(child render.RenderableNode) {
+	n.Children.ForEach(func(child report.Node) {
 		if child.ID != n.ID {
 			if summary, ok := MakeNodeSummary(child); ok {
-				summaries[child.Topology] = append(summaries[child.Topology], summary)
+				summaries[child.Topology] = append(summaries[child.Topology], summary.SummarizeMetrics())
```
Comment (tomwilkie) [inline render/detailed/node.go]: Nit, would prefer if this we de-nested a little:
```go
n.Children.ForEach(func(child report.Node) {
    if child.ID == n.ID {
        return
    }
    summary, ok := MakeNodeSummary(child)
    if !ok {
        return
    }
    summaries[child.Topology] = append(summaries[child.Topology], summary.SummarizeMetrics())
})
```

File: render/detailed/summary.go [hunk 5]
```diff
@@ -53,33 +64,58 @@ func MakeColumn(id string) Column {
...
 	Metrics      []MetricRow   `json:"metrics,omitempty"`
+	Adjacency    report.IDList `json:"adjacency,omitempty"`
 }
 
 // MakeNodeSummary summarizes a node, if possible.
-func MakeNodeSummary(n render.RenderableNode) (NodeSummary, bool) {
-	renderers := map[string]func(report.Node) NodeSummary{
+func MakeNodeSummary(n report.Node) (NodeSummary, bool) {
+	renderers := map[string]func(NodeSummary, report.Node) (NodeSummary, bool){
+		render.Pseudo:         pseudoNodeSummary,
```
Comment (tomwilkie) [inline render/detailed/summary.go]: This is a bit of a hack, but I can't think of a better way!

File: render/detailed/summary.go [hunk 6]
```diff
@@ -53,33 +64,58 @@ func MakeColumn(id string) Column {
...
 	Label        string        `json:"label"`
-	Linkable     bool          `json:"linkable"` // Whether this node can be linked-to
+	LabelMinor   string        `json:"label_minor"`
+	Rank         string        `json:"rank"`
+	Shape        string        `json:"shape,omitempty"`
+	Stack        bool          `json:"stack,omitempty"`
+	Linkable     bool          `json:"linkable,omitempty"` // Whether this node can be linked-to
+	Pseudo       bool          `json:"pseudo,omitempty"`
 	Metadata     []MetadataRow `json:"metadata,omitempty"`
 	DockerLabels []MetadataRow `json:"docker_labels,omitempty"`
```
Comment (tomwilkie) [inline render/detailed/summary.go]: It seems a shame a side effect of this change is that everything now has a docker labels.

I feel like we should change `Metadata` to be a `map[table name][]MetadataRow` to deal with this case.  Punt this for now though - maybe file a techdebt issue?

File: render/detailed/summary.go [hunk 7]
```diff
@@ -53,33 +64,58 @@ func MakeColumn(id string) Column {
...
+		Label:      n.Label,
+		LabelMinor: n.LabelMinor,
+		Rank:       n.Rank,
+		Shape:      n.Shape,
+		Stack:      n.Stack,
+		Linkable:   n.Linkable,
+		Adjacency:  n.Adjacency.Copy(),
 	}
 	for _, row := range n.Metadata {
 		result.Metadata = append(result.Metadata, row.Copy())
```
Comment (tomwilkie) [inline render/detailed/summary.go]: I feel like we should be using [`copy()`](https://golang.org/pkg/builtin/#copy) here

Comment (tomwilkie) [inline render/detailed/summary.go]: Ah ignore me.

File: render/detailed/summary.go [hunk 8]
```diff
@@ -93,52 +129,200 @@ func (n NodeSummary) Copy() NodeSummary {
...
 
-func baseNodeSummary(id, label string, linkable bool, nmd report.Node) NodeSummary {
-	return NodeSummary{
-		ID:           id,
-		Label:        label,
-		Linkable:     linkable,
-		Metadata:     NodeMetadata(nmd),
-		DockerLabels: NodeDockerLabels(nmd),
-		Metrics:      NodeMetrics(nmd),
+func baseNodeSummary(base NodeSummary, n report.Node) (NodeSummary, bool) {
```
Comment (tomwilkie) [inline render/detailed/summary.go]: Why does this need to return a bool?

File: render/detailed/summary.go [hunk 9]
```diff
@@ -93,52 +129,200 @@ func (n NodeSummary) Copy() NodeSummary {
...
+		base.Label = template.Label
+		base.LabelMinor = template.LabelMinor
+		base.Shape = template.Shape
+		return base, true
 	}
+
+	// try rendering it as an uncontained node
+	if strings.HasPrefix(n.ID, render.MakePseudoNodeID(render.UncontainedID)) {
+		base.Label = render.UncontainedMajor
+		base.Shape = Circle
```
Comment (tomwilkie) [inline render/detailed/summary.go]: Uncontained should be stacks of apps (squares), no?

File: render/detailed/summary.go [hunk 10]
```diff
@@ -93,52 +129,200 @@ func (n NodeSummary) Copy() NodeSummary {
...
+		return NodeSummary{}, false
+	}
+
+	imageNameWithoutVersion := render.ImageNameWithoutVersion(imageName)
+	base.Label = imageNameWithoutVersion
+	base.Rank = imageNameWithoutVersion
+	base.Shape = Hexagon
+	base.Stack = true
+
+	if base.Label == "<none>" {
```
Comment (tomwilkie) [inline render/detailed/summary.go]: Probably want this to be a constant.

File: render/detailed/summary.go [hunk 11]
```diff
@@ -93,52 +129,200 @@ func (n NodeSummary) Copy() NodeSummary {
...
+			base.Label = base.Label[:12]
+		}
+	}
+
+	if i, ok := n.Counters.Lookup(report.ContainerImage); ok {
+		if i == 1 {
+			base.LabelMinor = fmt.Sprintf("%d image", i)
+		} else {
+			base.LabelMinor = fmt.Sprintf("%d images", i)
+		}
```
Comment (tomwilkie) [inline render/detailed/summary.go]: When does this branch every get triggered?  When do we have groups with multiple container images?

File: render/detailed/summary.go [hunk 12]
```diff
@@ -53,33 +64,58 @@ func MakeColumn(id string) Column {
...
-	renderers := map[string]func(report.Node) NodeSummary{
+func MakeNodeSummary(n report.Node) (NodeSummary, bool) {
+	renderers := map[string]func(NodeSummary, report.Node) (NodeSummary, bool){
+		render.Pseudo:         pseudoNodeSummary,
 		report.Process:        processNodeSummary,
 		report.Container:      containerNodeSummary,
 		report.ContainerImage: containerImageNodeSummary,
 		report.Pod:            podNodeSummary,
+		report.Service:        serviceNodeSummary,
 		report.Host:           hostNodeSummary,
```
Comment (tomwilkie) [inline render/detailed/summary.go]: How are you dealing with containers-by-dns-name nodes?

File: render/expected/expected.go [hunk 13]
```diff
@@ -16,566 +14,110 @@ var (
...
-			),
-			EdgeMetadata: report.EdgeMetadata{},
-		},
-		unknownPseudoNode1ID:      unknownPseudoNode1(ServerProcessID),
-		unknownPseudoNode2ID:      unknownPseudoNode2(ServerProcessID),
-		render.IncomingInternetID: theIncomingInternetNode(ServerProcessID),
-		render.OutgoingInternetID: theOutgoingInternetNode,
-	}).Prune()
+	pseudo         = node(render.Pseudo)
+	endpoint       = node("") // TODO: endpoints don't have a topology for some reason? Not being tagged?
```
Comment (tomwilkie) [inline render/expected/expected.go]: Hmm they should.  Pls fix!

File: render/expected/expected.go [hunk 14]
```diff
@@ -16,566 +14,110 @@ var (
...
-			Shape:      cloud,
-			Node:       report.MakeNode().WithAdjacent(adjacent),
-			Children: render.MakeRenderableNodeSet(
-				RenderedEndpoints[RandomClientEndpointID],
-			),
-			EdgeMetadata: report.EdgeMetadata{
-				EgressPacketCount: newu64(60),
-				EgressByteCount:   newu64(600),
-			},
+	// Helper to make a report.node with some common options
```
Comment (tomwilkie) [inline render/expected/expected.go]: Like the new helper function!

But we seem to have lost all the children testing.  I think its important we keep that.

File: render/expected/expected.go [hunk 15]
```diff
@@ -16,566 +14,110 @@ var (
...
-
-	RenderedEndpoints = (render.RenderableNodes{
-		Client54001EndpointID: {
-			ID:    Client54001EndpointID,
-			Shape: circle,
-			Node:  report.MakeNode().WithAdjacent(ServerEndpointID),
-			EdgeMetadata: report.EdgeMetadata{
-				EgressPacketCount: newu64(10),
-				EgressByteCount:   newu64(100),
-			},
```
Comment (tomwilkie) [inline render/expected/expected.go]: How are we now testing EdgeMetadata?

File: render/expected/expected.go (hunk 14 above)
Comment (tomwilkie) [inline render/expected/expected.go]: Also, I think its important we have equivalent tests to these for a topolody of detailed nodes.  Maybe not all topologies, but at least the Process, Container and Host topo.

File: render/filters.go [hunk 16]
```diff
@@ -46,9 +47,7 @@ func ColorConnected(r Renderer) Renderer {
 
 			output := input.Copy()
 			for id := range connected {
-				output[id] = output[id].WithNode(report.MakeNodeWith(map[string]string{
-					IsConnected: "true",
-				}))
+				output[id] = output[id].WithLatests(map[string]string{IsConnected: "true"})
```
Comment (tomwilkie) [inline render/filters.go]: could be:
```go
output[id].WithLatest(IsConnected, mtime.Now(), "true")
```

File: render/id.go [hunk 17]
```diff
@@ -4,58 +4,7 @@ import (
...
-
-// MakeServiceID makes a service node ID for rendered nodes.
-func MakeServiceID(serviceID string) string {
-	return makeID("service", serviceID)
-}
-
-// MakeHostID makes a host node ID for rendered nodes.
-func MakeHostID(hostID string) string {
-	return makeID("host", hostID)
-}
```
Comment (tomwilkie) [inline render/id.go]: :+1:

File: render/mapping.go [hunk 18]
```diff
@@ -27,229 +26,80 @@ const (
...
-
-	node := NewRenderableNodeWith(id, label, minor, "", m)
-	node.ControlNode = m.ID
-	node.Shape = Hexagon
-	return RenderableNodes{id: node}
+	return NewDerivedPseudoNode(OutgoingInternetID, m)
 }
 
 // GetRenderableContainerName obtains a user-friendly container name, to render in the UI
 func GetRenderableContainerName(nmd report.Node) (string, bool) {
```
Comment (tomwilkie) [inline render/mapping.go]: Shouldn't this (and the constants) be moved to `detailed` package now?

File: render/mapping.go [hunk 19]
```diff
@@ -259,10 +109,9 @@ func MapEndpoint2IP(m RenderableNode, local report.Networks) RenderableNodes {
 	// So we need to emit two nodes, for two different cases.
 	id := report.MakeScopedEndpointNodeID(scope, addr, "")
 	idWithPort := report.MakeScopedEndpointNodeID(scope, addr, port)
-	m = m.WithParents(report.EmptySets)
-	return RenderableNodes{
-		id:         NewRenderableNodeWith(id, "", "", "", m),
-		idWithPort: NewRenderableNodeWith(idWithPort, "", "", "", m),
+	return report.Nodes{
+		id:         NewDerivedNode(id, m).WithTopology("IP"),
```
Comment (tomwilkie) [inline render/mapping.go]: `"IP"` constant (to go with `"Pseudo"`)?

File: render/mapping.go [hunk 20]
```diff
@@ -365,26 +220,51 @@ func MapEndpoint2Pseudo(n RenderableNode, local report.Networks) RenderableNodes
...
+	return report.Nodes{id: node}
+}
+
+// MapTo re-indexes a set of nodes by some key from Latest, changing the topology
+// TODO(paulbellamy): not sure this works, or is general enough
+func MapTo(topology, key string) func(report.Node, report.Networks) report.Nodes {
+	return func(n report.Node, local report.Networks) report.Nodes {
+		// Pass through pseudo nodes
+		if n.Topology == Pseudo {
+			fmt.Printf("[DEBUG] MapTo skipping %s %s", n.Topology, n.ID)
```
Comment (tomwilkie) [inline render/mapping.go]: Debug line needs removing.

File: render/mapping.go [hunk 21]
```diff
@@ -365,26 +220,51 @@ func MapEndpoint2Pseudo(n RenderableNode, local report.Networks) RenderableNodes
...
+			return report.Nodes{n.ID: n}
+		}
+
+		// Throw away any nodes not having this key
+		entry, ok := n.Latest.LookupEntry(key)
+		if !ok {
+			fmt.Printf("[DEBUG] MapTo skipping %s %s: no key %s", n.Topology, n.ID, key)
```
Comment (tomwilkie) [inline render/mapping.go]: Debug line needs removing.

File: render/mapping.go [hunk 22]
```diff
@@ -365,26 +220,51 @@ func MapEndpoint2Pseudo(n RenderableNode, local report.Networks) RenderableNodes
...
+	id := report.MakeProcessNodeID(report.ExtractHostID(n), pid.Value)
+	node := NewDerivedNode(id, n).WithTopology(report.Process)
+	node.Latest = node.Latest.Set(process.PID, pid.Timestamp, pid.Value)
+	node.Counters = node.Counters.Add(n.Topology, 1)
+	return report.Nodes{id: node}
+}
+
+// MapTo re-indexes a set of nodes by some key from Latest, changing the topology
+// TODO(paulbellamy): not sure this works, or is general enough
+func MapTo(topology, key string) func(report.Node, report.Networks) report.Nodes {
```
Comment (tomwilkie) [inline render/mapping.go]: Not actually used anywhere?

File: report/latest_map.go [hunk 23]
```diff
@@ -96,6 +96,18 @@ func (m LatestMap) Lookup(key string) (string, bool) {
 	return value.(LatestEntry).Value, true
 }
 
+// LookupEntry returns the raw entry for the given key.
+func (m LatestMap) LookupEntry(key string) (LatestEntry, bool) {
```
Comment (tomwilkie) [inline report/latest_map.go]: Instead of exposing LatestEntry, can we return `(string, time.Time, bool)`?

File: report/node_set.go [hunk 24]
```diff
@@ -13,35 +13,35 @@ import (
...
+// Add adds the nodes to the NodeSet. Add is the only valid way to grow a
+// NodeSet. Add returns the NodeSet to enable chaining.
+func (n NodeSet) Add(nodes ...Node) NodeSet {
 	result := n.psMap
 	if result == nil {
 		result = ps.NewMap()
 	}
 	for _, node := range nodes {
-		result = result.Set(fmt.Sprintf("%s|%s", node.Topology, node.ID), node)
+		result = result.Set(node.ID, node)
```
Comment (tomwilkie) [inline report/node_set.go]: \o/

We could also change a Topology to be a NodeSet... in a future PR.

File: test/fixture/report_fixture.go [hunk 25]
```diff
@@ -121,7 +121,7 @@ var (
 				// Node is arbitrary. We're free to put only precisely what we
 				// care to test into the fixture. Just be sure to include the bits
 				// that the mapping funcs extract :)
-				Client54001NodeID: report.MakeNode().WithLatests(map[string]string{
+				Client54001NodeID: report.MakeNode().WithID(Client54001NodeID).WithLatests(map[string]string{
```
Comment (tomwilkie) [inline test/fixture/report_fixture.go]: :-(

**Comment by tomwilkie** [comment]
Like this a lot.  Some stuff in experimental needs fixing, and I think we need to add some testing for the final, fully rendered NodeSummaries.

Also I think we need to address #836 ASAP.  And we can potentially experiment with not merging in existing metadata when we do a derived node.

**Comment by tomwilkie** [comment]
Have run against example app, seems to do all the right things.

File: test/fixture/report_fixture.go (hunk 25 above)
Comment (paulbellamy) [inline test/fixture/report_fixture.go]: yeeeap

File: render/mapping.go (hunk 22 above)
Comment (paulbellamy) [inline render/mapping.go]: Yeah. had put it in as an experiment in generalizing the MapX2Y functions. Was curious if you think it's worth pursuing, or bailing on that.

File: report/latest_map.go (hunk 23 above)
Comment (paulbellamy) [inline report/latest_map.go]: Sure. Actually I don't know if we use this in the end, or can remove it.

File: report/node_set.go (hunk 24 above)
Comment (paulbellamy) [inline report/node_set.go]: :+1:

File: render/mapping.go (hunk 22 above)
Comment (tomwilkie) [inline render/mapping.go]: I think with this change something like this should be possible, but I'd stash it for a future PR.

File: report/latest_map.go (hunk 23 above)
Comment (tomwilkie) [inline report/latest_map.go]: Oh I like the usages of it (in mapping.go).

File: render/detailed/summary.go (hunk 6 above)
Comment (paulbellamy) [inline render/detailed/summary.go]: raised

File: render/detailed/summary.go (hunk 11 above)
Comment (paulbellamy) [inline render/detailed/summary.go]: At the moment nowhere. Felt like it should be there for completeness with the other renderers.

File: render/detailed/summary.go (hunk 12 above)
Comment (paulbellamy) [inline render/detailed/summary.go]: https://github.com/weaveworks/scope/pull/1204/files#diff-f104a6024333d20eb4e87d1dfa366ebfR77
and
https://github.com/weaveworks/scope/pull/1204/files#diff-d381aa4876e4f400a35de540793f8977R207

Bit of a hack, as it's not really extensible to grouping by other keys. But I figure we can work that out when we do the dsl.

Comment (tomwilkie) [inline render/detailed/summary.go]: I wonder whether we should have another 'pseudo-topology' (called "Grouped") where we do some simple heuristics, like pulling a label from metadata.  Then use this for grouped topologies that don't have a 'real' node - like process-by-name and container-by-hostname.

Comment (paulbellamy) [inline render/detailed/summary.go]: or do a prefix like `pseudo:...`, do `group:...`. I had thought about that. Could pass the label info in the ID (eugh, but...), like `group:<original topology>:<key>:<value>`. Or we could do `group:<value>` and set the label value as a special key in the metadata. either way. I figure its best to delay deciding on that until we have a driver (dsl).

File: render/expected/expected.go (hunk 14 above)
Comment (paulbellamy) [inline render/expected/expected.go]: Will add those

File: render/expected/expected.go (hunk 15 above)
Comment (paulbellamy) [inline render/expected/expected.go]: Same way we were before (we're not).

**Comment by tomwilkie** [comment]
LGTM on this.

**Comment by paulbellamy** [comment]
Rebased on 3744f475613f86a9506b1509e56339dec8ba9205

**Pull request merged**